include/oneapi/dpl/internal/dynamic_selection_impl/concurrent_queue.h

```diff
@@ -0,0 +1,76 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) 2023 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+
+#ifndef _ONEDPL_CONCURRENT_QUEUE_H
+#define _ONEDPL_CONCURRENT_QUEUE_H
+
+#include <queue>
+#include <mutex>
+#include <list>
+#include <condition_variable>
+
+namespace oneapi {
+namespace dpl{
+namespace experimental{
+namespace util{
+    template <typename T>
+    class concurrent_queue
+    {
+    public:
+
+        void pop(T& item)
+        {
+            std::unique_lock<std::mutex> mlock(mutex_);
+            while (queue_.empty())
+            {
+                cond_.wait(mlock);
+            }
+            item = queue_.front();
+            queue_.pop();
+        }
+
+        void push(T item)
+        {
+            std::unique_lock<std::mutex> mlock(mutex_);
+            queue_.push(item);
+            mlock.unlock();
+            cond_.notify_one();
+        }
+
+        void pop_all(std::list<T>& item_list)
+        {
+            std::unique_lock<std::mutex> mlock(mutex_);
+            while(!queue_.empty()){
+                auto item = queue_.front();
+                queue_.pop();
+                item_list.push_back(item);
+
+            }
+        }
```

Comment:
This method could be made more efficient, but would need API adjustments.
Remember that `std::queue` is a "container adaptor", i.e. it builds atop of another container (`std::deque` by default but could also be a list or a vector). So, if the `concurrent_queue` is also made a container adaptor, for `pop_all` we might just `std::move` the underlying container to the caller, instead of popping & pushing individual elements.